Lib/unittest/test/testmock/testmock.py

```diff
@@ -1847,6 +1847,39 @@ def foo(a, b):
         self.assertRaises(TypeError, mock.child, 1)
         self.assertEqual(mock.mock_calls, [call.child(1, 2)])
 
+    def test_isinstance_under_settrace(self):
+        # bpo-36593 : __class__ is not set for a class that has __class__
+        # property defined when it's used with sys.settrace(trace) set.
+        # Delete the module to force reimport with tracing function set
+        # restore the old reference later since there are other tests that are
+        # dependent on unittest.mock.patch. In testpatch.PatchTest
+        # test_patch_dict_test_prefix and test_patch_test_prefix not restoring
+        # causes the objects patched to go out of sync
+
+        old_patch = unittest.mock.patch
+        with patch.dict('sys.modules'):
+            del sys.modules['unittest.mock']
+
+            def trace(frame, event, arg):
+                return trace
+
+            sys.settrace(trace)
+            self.addCleanup(sys.settrace, None)
+
+            from unittest.mock import (
+                Mock, MagicMock, NonCallableMock, NonCallableMagicMock
+            )
+
+            mocks = [
+                Mock, MagicMock, NonCallableMock, NonCallableMagicMock
+            ]
+
+            for mock in mocks:
+                obj = mock(spec=Something)
+                self.assertIsInstance(obj, Something)
+
+        unittest.mock.patch = old_patch
```

Comment:
Use `self.addCleanup(` to make sure the restoring is made even if something fails in the test.

Reply:
I tried using `__settattr__` right after old_patch assignment as below so that this is covered but this fails to restore update the patch object correctly. I think maybe the older reference to `unittest.patch` is added during `addCleanup` and since I re-import `unittest.mock` inside the context manager. 

I can add it at the end to update the correct reference but that would mean cleanup is not registered despite the test case failure.

```python
old_patch = unittest.mock.patch
self.addCleanup(setattr, unittest.mock, 'patch', old_patch)
```

So I have used a `_cleanup` function so that during function execution it refers to the updated `unittest.mock` object. Added a comment on the same. Verified the same by causing custom exception inside context manager and making sure the other tests don't fail.

Reply:
You could just use a lambda for creating the closure, right? (Or your current approach :) )

Reply:
yes, I was thinking along `self.addCleanup(lambda patch: unittest.mock.patch = patch, old_patch)` only to remind myself assignment doesn't work inside lambdas. Since this uses `setattr` I have used your suggestion. Thanks for the tip.

Reply:
Thank you @tirkarthi for iterating over this nit and for you thorough analysis of the problem